enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CartesianProductRegisterPipe.scala

```diff
@@ -19,18 +19,26 @@
  */
 package org.neo4j.cypher.internal.compatibility.v3_3.runtime.interpreted.pipes
 
-import org.neo4j.cypher.internal.compatibility.v3_3.runtime.ExecutionContext
-import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.{Pipe, PipeWithSource, QueryState}
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.interpreted.PrimitiveExecutionContext
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.{Pipe, QueryState}
 import org.neo4j.cypher.internal.compatibility.v3_3.runtime.planDescription.Id
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.{ExecutionContext, PipelineInformation}
 
-case class SemiApplyRegisterPipe(source: Pipe, inner: Pipe, negated: Boolean)
-                                (val id: Id = new Id)
-  extends PipeWithSource(source) {
-  def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
-    input filter { (inputContext) =>
-      val innerState = state.withInitialContext(inputContext)
-      val innerResults = inner.createResults(innerState)
-      if (negated) innerResults.isEmpty else innerResults.nonEmpty
+case class CartesianProductRegisterPipe(lhs: Pipe, rhs: Pipe,
+                                        lhsLongCount: Int, lhsRefCount: Int,
+                                        pipelineInformation: PipelineInformation)
+                                       (val id: Id = new Id) extends Pipe {
+
+  protected def internalCreateResults(state: QueryState): Iterator[ExecutionContext] = {
+    lhs.createResults(state) flatMap {
+      lhsCtx =>
+        rhs.createResults(state) map {
+          rhsCtx =>
+            val context = PrimitiveExecutionContext(pipelineInformation)
+            lhsCtx.copyTo(context)
+            rhsCtx.copyTo(context, lhsLongCount, lhsRefCount)
```

Comment:
The lhsLongCount and lhsRefCount would be known by the lhsCtx already, so why do we need to pass it in to the CartesianProductRegisterPipe? Would it not be better to use the CartesianProductPipe pattern of calling `lhsCtx mergeWith rhsCtx` and have the PrimitiveExecutionContext implementation of `mergeWith` understand how to merge two execution contexts like this? Then we could also remove the CartesianProductRegisterPipe class entirely.

Reply:
One issue I see with my idea is that the Cartesian Product pipeline is provided at register allocation time. But it is strictly a concatenation of left and right hand side pipelines, and we are doing that with mergeWith also (at the execution context level). I still feel there are ways to simplify this.